src/software/new_geom/util/intersects_test.cpp

```diff
@@ -0,0 +1,414 @@
+#include "software/new_geom/util/intersects.h"
+
+#include <gtest/gtest.h>
+
+#include "software/new_geom/rectangle.h"
+#include "software/new_geom/triangle.h"
+
+TEST(IntersectsTest, polygon_segment_intersecting_on_edge)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Segment s({-7, 0}, {-4, 1});
+    EXPECT_TRUE(intersectsNew(p, s));
+    EXPECT_TRUE(intersectsNew(s, p));
+}
+
+TEST(IntersectsTest, polygon_segment_intersecting_on_vertex)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Segment s({-2, 0}, {0, 2});
+    EXPECT_TRUE(intersectsNew(p, s));
+    EXPECT_TRUE(intersectsNew(s, p));
+}
+
+TEST(IntersectsTest, polygon_far_from_segment_not_intersecting)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Segment s({100000, 20000}, {-5000, 200});
+    EXPECT_FALSE(intersectsNew(p, s));
+    EXPECT_FALSE(intersectsNew(s, p));
+}
+
+TEST(IntersectsTest, polygon_near_segment_not_intersecting)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Segment s({-6, 2.5}, {-1, 1.5});
+    EXPECT_FALSE(intersectsNew(p, s));
+    EXPECT_FALSE(intersectsNew(s, p));
+}
+
+TEST(IntersectsTest, segment_overlapping_polygon_edge)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Segment s({2, -2.5}, {8, -9.5});
+    EXPECT_TRUE(intersectsNew(p, s));
+    EXPECT_TRUE(intersectsNew(s, p));
+}
+
+TEST(IntersectsTest, ray_inside_polygon)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Ray r({2, -2}, Angle::fromDegrees(20));
+    EXPECT_TRUE(intersectsNew(p, r));
+    EXPECT_TRUE(intersectsNew(r, p));
+}
+
+TEST(IntersectsTest, ray_outside_polygon_intersecting)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Ray r({-2, -6}, Angle::fromDegrees(100));
+    EXPECT_TRUE(intersectsNew(p, r));
+    EXPECT_TRUE(intersectsNew(r, p));
+}
+
+TEST(IntersectsTest, ray_outside_polygon_not_intersecting)
+{
+    Polygon p({{-6, 2}, {-1, 1}, {10, 7}, {5, -6}, {-5, -3}});
+    Ray r({-2, -6}, Angle::fromDegrees(180));
+    EXPECT_FALSE(intersectsNew(p, r));
+    EXPECT_FALSE(intersectsNew(r, p));
+}
+
+TEST(IntersectsTest, circle_tangent_to_triangle_edge)
+{
+    Triangle t({-5, 0}, {5, 0}, {2, 5});
+    Circle c({0, -1}, 1);
+    EXPECT_FALSE(intersectsNew(t, c));
+    EXPECT_FALSE(intersectsNew(c, t));
+}
+
+TEST(IntersectsTest, circle_completely_inside_triangle)
```

Comment:
Should we consider objects in side another to be intersecting? In some way this seems to overlap with `contains()`. I'm not sure if any behavior currently depends on this, but we should all make sure we're on the same page. I think the fundamental assumption here is whether or not 2D shapes (Circles and Polygons) actually occupy their enclosed area, or are simply outlines. I think there are cases where both are useful.

We should also make sure this behavior matches the equivalent `intersection` functions.

For example, a circle completely contained inside a polygon will return in `intersects` returning `true`. However if I called the `intersection` function (assuming on existed for these shapes) to actually see where the shapes intersect, I would expect it to return `std::nullopt` because their boundaries do not intersect. So the behavior from `intersects` and intersections` are inconsistent based on what I'd expect.

This may be a conversation worth moving to Slack to get more people involved.

Reply:
In terms of what's implemented, there is no overlap with `contains` at the moment since there are no `contains` functions dealing with two `Shape`s.

The behaviours of `intersects` and `intersection` functions seem consistent with Boost geometry (https://www.boost.org/doc/libs/1_72_0/libs/geometry/doc/html/geometry/reference/algorithms/intersection/intersection_3.html) where `intersects` checks if two geometries have at least one intersection, and `intersection` returns the intersection based on user specified type (`Polygon`, `LineString` or `Point`). In a case where a 2D shape is completely contained with in another 2D shape, the Boost `intersects` would return true but `intersection` with output specified as `Point` would not return any results. I would prefer to stay consistent with this interface.

My suggestion is to keep `intersects` and `intersection` as is, but define a new `crosses` / `overlaps` function if we want to say for sure that there are intersecting points between geometries(https://en.wikipedia.org/wiki/DE-9IM#/media/File:TopologicSpatialRelarions2.png).

Reply:
So this is the definition of intersection from wikipedia: "In geometry, an intersection is a point, line, or curve common to two or more objects (such as lines, curves, planes, and surfaces)."l[0]. This definition fits with what Jason is saying here, and in general I agree. 

[0]: https://en.wikipedia.org/wiki/Intersection_%28Euclidean_geometry%29

Reply:
Ok that's fair. Can that be documented somewhere either at the top of each file (intersects and intersection) or could we specify that in the javadoc comments? ie. that `intersects` returns true if the shapes overlap at all (as per the definitions above)? Just so it's clear to whoever goes to use these later.

Reply:
Left documentation at the top of `intersects.h` and `intersection.h`.